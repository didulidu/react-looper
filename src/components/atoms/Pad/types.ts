import { Pad } from "../../../model/Pad";

export type PadProps = {
  pad: Pad;
  onPress: (id: string) => void;
};

export type StyledPadProps = { isPressed: boolean; isPlaying: boolean };
