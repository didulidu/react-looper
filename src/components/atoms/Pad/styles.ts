import styled from "styled-components";
import { StyledPadProps } from "./types";

export const StyledPad = styled.div`
  width: 200px;
  height: 200px;
  border: 10px outset dimgrey;
  border-radius: 35px;
  box-shadow: rgb(38, 57, 77) 0px 20px 30px -10px;
  margin: 10px;
  background-color: ${(props: StyledPadProps) =>
    props.isPressed
      ? props.isPlaying
        ? "salmon"
        : "lightblue"
      : "lightgreen"};
  display: flex;
  justify-content: center;
  align-items: center;

  @media only screen and (max-width: 900px) {
    width: 100px;
    height: 100px;
  }
`;

export const iconDimension = {
  width: "50%",
  height: "50%",
};
