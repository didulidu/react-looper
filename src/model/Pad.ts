export type Pad = {
  id: string;
  active: boolean;
  isPlaying: boolean;
  audio: HTMLAudioElement;
};
