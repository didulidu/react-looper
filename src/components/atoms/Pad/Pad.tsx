import { FC, memo } from "react";
import { ReactComponent as Playing } from "../../../assets/icons/playing.svg";
import { ReactComponent as Play } from "../../../assets/icons/play.svg";
import { ReactComponent as WillPlay } from "../../../assets/icons/will_play.svg";
import { PadProps } from "./types";
import { StyledPad } from "./styles";
import useAudio from "../../../hooks/useAudio";

const Pad: FC<PadProps> = ({ pad, onPress }) => {
  useAudio(pad);
  const Icon = pad.active ? (pad.isPlaying ? Playing : WillPlay) : Play;

  return (
    <>
      <StyledPad
        onClick={() => onPress(pad.id)}
        isPressed={pad.active}
        isPlaying={pad.isPlaying}
      >
        <Icon width="100px" height="100px" />
      </StyledPad>
    </>
  );
};

export default memo(Pad, (prevProps, nextProps) => {
  return (
    prevProps.pad.active === nextProps.pad.active &&
    prevProps.pad.isPlaying === nextProps.pad.isPlaying
  );
});
