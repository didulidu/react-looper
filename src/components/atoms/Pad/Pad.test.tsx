import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Pad from "./Pad";

describe("Pad test", () => {
  test("Looper renders correctly", () => {
    const { container } = render(
      <Pad
        onPress={jest.fn()}
        pad={{
          id: "1",
          isPlaying: false,
          active: false,
          audio: new Audio("./loop1.mp3"),
        }}
      />
    );
    expect(screen.getByLabelText("Pad")).toHaveStyle({
      "background-color": "lightgreen",
    });
    expect(container).toMatchSnapshot();
  });
  test("Pad click", () => {
    const onPressMock = jest.fn();
    render(
      <Pad
        onPress={onPressMock}
        pad={{
          id: "1",
          isPlaying: false,
          active: false,
          audio: new Audio("./loop1.mp3"),
        }}
      />
    );
    userEvent.click(screen.getByLabelText("Pad"));
    expect(onPressMock).toBeCalledTimes(1);
  });
  test("Pad active", () => {
    const { container } = render(
      <Pad
        onPress={jest.fn()}
        pad={{
          id: "1",
          isPlaying: false,
          active: true,
          audio: new Audio("./loop1.mp3"),
        }}
      />
    );
    expect(screen.getByLabelText("Pad")).toHaveStyle({
      "background-color": "lightblue",
    });
    expect(container).toMatchSnapshot();
  });

  test("Pad playing", () => {
    const { container } = render(
      <Pad
        onPress={jest.fn()}
        pad={{
          id: "1",
          isPlaying: true,
          active: true,
          audio: new Audio("./loop1.mp3"),
        }}
      />
    );
    expect(screen.getByLabelText("Pad")).toHaveStyle({
      "background-color": "salmon",
    });
    expect(container).toMatchSnapshot();
  });
});
