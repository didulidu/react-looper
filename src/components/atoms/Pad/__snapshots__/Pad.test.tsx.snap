// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pad test Looper renders correctly 1`] = `
<div>
  <div
    aria-label="Pad"
    class="sc-bdvvtL fngWLE"
  >
    <svg
      height="50%"
      width="50%"
    >
      play.svg
    </svg>
  </div>
</div>
`;

exports[`Pad test Pad active 1`] = `
<div>
  <div
    aria-label="Pad"
    class="sc-bdvvtL bMQvHd"
  >
    <svg
      height="50%"
      width="50%"
    >
      will_play.svg
    </svg>
  </div>
</div>
`;

exports[`Pad test Pad playing 1`] = `
<div>
  <div
    aria-label="Pad"
    class="sc-bdvvtL gISjDH"
  >
    <svg
      height="50%"
      width="50%"
    >
      playing.svg
    </svg>
  </div>
</div>
`;
